### ISSUE TITLE
clientFromConnectionString causes process to hang

### ISSUE THREAD

**Issue by gulipek95** [Azure/azure-iot-sdk-node#285]
# Context
- **OS and version used:** Windows 10
- **Node.js version:** v8.9.4
- **npm version:** 5.6.0

# Description of the issue:
This issue came up when we were adding unit tests that used the azure-iot-device-amqp module, and the tests never finish executing. Everything else is cleaned, connections closed, but the client instance creates some hidden processes. To make sure it was the instantiation, I created a simple .js file with the content below, and after the connection is closed, it continues to run (max I waited was 5 min, but I think that is plenty enough) until I Ctrl+C.

# Code sample exhibiting the issue (this is the whole js file):
const azureClient = require('azure-iot-device-amqp');
const connectionString = <my valid connection string>;

const client = azureClient.clientFromConnectionString(connectionString);

client.open((error, result) => {
  console.log(`open error is: ${error}`);
  console.log(`open result is: ${result}`);

  console.log('about to close the client');
  client.close((err, res) => {
    console.log(`close error is: ${err}`);
    console.log(`close result is: ${res}`);

    console.log('closed the client');
  });
});

# Console log of the issue:
open error is: null
open result is: [object Object]
about to close the client
close error is: null
close result is: [object Object]
closed the client

.. and the code hangs forever, doesn't exist.

**Comment by pierreca** [comment]
Thanks for reporting this. I believe it’s related to the timer that is started in the SharedAccessKeyAuthenticationProvider class. One workaround would be to use x509 certificates instead, while we validate this hypothesis and fix the bug.

**Comment by gulipek95** [comment]
You're welcome - for x509 certificate, where can I use/set it to be the cert?

**Comment by anthonyvercolano** [comment]
simple_sample_device_x509 in the azure-iot-sdk-node/device/samples directory demonstrates.

In the build/tools directory there is a tool called create_device_certs.js that will create a self signed cert/private keythat you could use.  The C sdk has a more sophisticated tool chain for creating a cert chain if you want.

https://github.com/Azure/azure-iot-sdk-c/blob/master/tools/CACertificates/CACertificateOverview.md


Or, you could just add a process.exit() and that would probably work too.

I'll follow up on Monday with investigation results.

**Comment by gulipek95** [comment]
Cheers! Process.exit is how I am handling it right now - will wait for the updates, thank you.

**Comment by anthonyvercolano** [comment]
fyi:  This was indeed a problem with the SharedAccessKeyAuthenticationProvider class.  The fix for this creates some interesting interface issues.  I'm likely to want to wait till Pierre returns in this case.  This could delay a fix until mid June.  Will you be ok with the process.exit solution until then?

**Comment by gulipek95** [comment]
Hi Anthony, 

I am ok with using process.exit until it is fixed, it is just for unit tests. My greater concern is my actual use case of the module.

I was wondering if you could give me a bit more context on what exactly causes the problem. In the regular use case I instantiate an azureClient for each connect request (currently it is a 1 to 1 relationship, but in the future it will be many to many) my server receives. I create a notification channel back to whoever made the connect request, and I keep up the azureClient connection while the notification channel is alive. 

If it goes down I take down the azureClient (by calling .close() ) and then call;
delete this._azureClient; (in the class that manages these connections)

Then I receive the request to connect again, and instantiate a new azureClient. If for some reason this cycle happens frequently, would it cause a leak/multiple hanging processes in the background? 

In other words, is the problem valid only when the connection is closed too fast, or it exists regardless?

Thank you for your time!

**Comment by anthonyvercolano** [comment]
tldr: Exists regardless.

Actually you can probably get into this issue without even the open or the close.  Just calling the factory method is enough.  There is an authentication "provider" interface in the code (you never see this yourself) that I believe should change to have an open/close method added.  Currently the constructor for this provider is starting a timer that regenerates authentication information at a set interval.  The code should be changed to not start the timer until the device is actually opened and stop the timer when the device is closed.

This could be construed as a breaking change to an api that developers would not normally be expected to invoke ever.  Just want to be clear this.  I expect to hear from Pierre next week.

**Comment by gulipek95** [comment]
Thank you for your response. For the open/close method that the provider will invoke (behind the scenes), will they be connected to the open/close API that the Client module expose? 

I really appreciate both your and Pierre's immediate attention on the matter!

**Comment by anthonyvercolano** [comment]
Assuming others like the idea, the providers open/close methods would be invoked by the clients open/close methods.

**Comment by pierreca** [comment]
This was fixed with #336 and released with v1.7.1 of the packages.
@gulipek95 thanks for your patience with this!

*Please help us improve by filling out this 2-minute [customer satisfaction survey](https://www.surveymonkey.com/r/Q5PLZ3V)*
